### PR TITLE
US-3381 Added several tests to cover implemented calls

### PR DESCRIPTION
**Pull request by pfoster-youitv**
## Change list

Added new io.appium.java_client.YouiEngine package to enable Appium to automate interactions with apps built using [You.i Engine](http://www.youi.tv/).

Package includes:

 - ./YouiEngine/YouiEngineDriver.java
 - ./YouiEngine/YouiEngineElement.java
 - ./YouiEngine/internal/JsonToYouiEngineElementConverter.java
 - ./remote/YouiEngineCapabilityType.java

Under the test side of the package:

 - ./YouiEngine/SanityTest.java - based on the Java Client tutorial
 - ./YouiEngine/YouiEngineAppiumSampleTest.java - self contained demo
 - ./YouiEngine/util/AppiumTest.java - based on the Java Client tutorial
 - ./YouiEngine/util/TestUtiliy.java - common utility methods for tests
 - ./YouiEngineAppiumSample.app.zip - target app for above scripts when `isAndroid = false`
 - ./YouiEngineAppiumSample-debug.apk - target app for above scripts when `isAndroid = true`

No changes made to existing files, only additions to specific to You.i Engine.

Google coding standards were applied with all error-level violations being corrected.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Details

The only difference between writing a test for You.i Engine and any other Appium platform is the use of `platformName = "youiengine"` and the pressence of `youiEngineAppPlatform` and `youiEngineAppAddress`. The `youiEngineAppPlatform` should be set to the device platform; `android` or `ios`. The `youiEngineAppAddress` must be set to the device's IP address. 

Review the `YouiEngineAppiumSampleTest.java` for one way to write tests against a Youi Engine app using this Appium client. Adjust the capabilities found within this class to suit the environment the test will be run in, including the target device. The same test class can be used for Android or iOS by toggling the `isAndroid` boolean accordingly.

#### Supported Platforms

Android and iOS are currently supported. 

#### Requirements

This client requires the `appium-youiengine-driver`.

#### Available Interactions
Following is a list of available interactions with notes on which platforms they are supported with.

##### Common

> closeApp 
> findElement* 
> findElements* 
> getAttribute** 
> getContext
> getContextHandles 
> getLogs
> getLogTypes
> getScreenShotAs 
> getSessionId 
> getText 
> installApp
> launchApp 
> removeApp 
> runAppInBackground 
> sendKeys\***

\* currently supports `className`, `id` and `name` `By` strategies .
\** currently supports `name` attribute only.
\*** special key code support not available.
##### Exclusive to Android

> isAppInstalled 
> isLocked
> lock
> lockFor*
> toggleLocationServices
> unlock

\* this implementation takes in the number of seconds to lock the device for.
##### Exclusive to iOS

> lockFor* 

\* this implementation takes in the number of seconds to lock the device for.

NOTE: Unlock calls will only work when unlocking is set to 'swipe to unlock'.

### CODE REVIEW

**Comment by pfoster-youitv** [comment]
The above commit message is the current one for when we'll post officially to the Appium repo... feedback on it is appreciated too.

File: src/test/java/io/appium/java_client/YouiEngine/SanityTest.java [hunk 1]
```diff
@@ -203,6 +206,69 @@ public void toggleLocationServicesTest() throws Exception {
...
+                    + "supplied device.");
+        }
+    }
+
+    // Confirm we can toggle the Android device's wifi setting.
+    @org.junit.Test
+    public void toggleWifiTest() throws Exception {
+        try {
+            driver.toggleWiFi(); // off
+            utils.delayInSeconds(5);
```
Comment (daverooneyca) [inline src/test/java/io/appium/java_client/YouiEngine/SanityTest.java]: Are you able to assert that wifi has been turned off at this point, or would an exception be thrown by the driver?

File: src/test/java/io/appium/java_client/YouiEngine/SanityTest.java [hunk 2]
```diff
@@ -203,6 +206,69 @@ public void toggleLocationServicesTest() throws Exception {
...
+                    + "supplied device.");
+        }
+    }
+
+    // Confirm we can toggle the Android device's data setting.
+    @org.junit.Test
+    public void toggleDataTest() throws Exception {
+        try {
+            driver.toggleData(); // off
+            utils.delayInSeconds(5);
```
Comment (daverooneyca) [inline src/test/java/io/appium/java_client/YouiEngine/SanityTest.java]: Same question - can you assert that data has been turned off?

**Comment by pfoster-youitv** [comment]
After a bit of testing, seems the following commands end up failing to be sent from the client because they are being evaluated as 'not supported' deep down in the client. We need a way to prevent this or bypass this check.

#### iOS
> mobileShake 
#### Android
> toggleData
> toggleFlightMode
> toggleWifi

**Comment by pfoster-youitv** [comment]
Updated again. Added some overrides that will assert that a particular By strategy is not supported in YouiEngine. I'd prefer to throw an exception, but seems that's not allowed with an override? I need to look into this some more.

**Comment by HeloiseRD** [comment]
+1 I've used it and it seems to work fine.
Doesn't handle exceptions properly so I created a Jira for that - US-3492

**Pull request merged**